### PR TITLE
gh-126742: allow to use non-UTF8 exception messages

### PR DESCRIPTION
**Pull request by picnixz**
* Issue: gh-126742

### CODE REVIEW

File: Python/errors.c
Comment (ZeroIntensity) [inline Python/errors.c]: I think this should be `surrogateescape`, not `strict`:
```suggestion
    PyObject *value = PyUnicode_DecodeLocale(string, "surrogateescape");
```

File: Python/errors.c
Comment (ZeroIntensity) [inline Python/errors.c]: It might be worth just turning this into an inline function or macro.

**Comment by ZeroIntensity** [comment]
Technically, since this is a bugfix, let's wait until #126555 has landed before we merge this, and then address all the uses of `dlerror()` in this PR.

File: Python/errors.c
Comment (picnixz) [reply Python/errors.c]: Both are acceptable and I think it's preferrable to use a strict decoding since a message that is not acceptable by the current locale feels wrong. In addition, using surrogateescape means that for re-encoding the message you need to also pass the "surrogateescape" handler.

File: Python/errors.c
Comment (picnixz) [reply Python/errors.c]: Others functions are implemented without inlining and without macros so I think it's fine to keep it that way here.

**Comment by picnixz** [comment]
> then address all the uses of dlerror() in this PR.

I actually thought of doing it two PRs.

**Comment by ZeroIntensity** [comment]
Eh, I would rather not. Let's not overwhelm someone with thread states and the private API on their first PR :)

File: Python/errors.c
Comment (picnixz) [reply Python/errors.c]: Ok, looking around, other usages in the context of error rendering use the "surrogateescape" handler, so I've decided to use it as well.

**Comment by ZeroIntensity** [comment]
I've removed backport labels, per what Petr said in the issue. I think he's right--let's just focus on main with this.

**Comment by ZeroIntensity** [comment]
I like that idea! But we're exposing this as a public API--how often might users need this? (I've never needed to mess with locale in any of my C API usage, but to be fair I'm typically not calling APIs that return strings.)

**Comment by picnixz** [comment]
> Hmm, just had an idea: if we add a “char* in locale encoding” format to 

I'm not very fond of adding a case to additionally check when there's not a lot of use cases. It shouldn't affect performances much (but this needs verification) but this would complicate an already complicate function I think (and people need to remember this new format as well).

**Comment by picnixz** [comment]
> No no, I don't think he meant a variation of PyUnicode_FromFormat that takes a locale, he meant a format value %whatever that automatically calls PyUnicode_DecodeLocale or something like that.

And I understood. I don't think a new code should be added unless there is a real need.

**Comment by ZeroIntensity** [comment]
> And I understood.

Ah! I couldn't tell, sorry.

A code search for `PyUnicode_DecodeLocale` is showing not too much real world usage--maybe that's a bad thing? Depending on how common it is for a non-UTF8 locale to break things, maybe it would be a good idea to add a format character just so people start handling it.

**Comment by encukou** [comment]
It could also show that we're mainly handling errors from the OS at one point -- the OSError constructor. That would be a good thing :)

Anyway, for this PR: @picnixz, do you want to add calls to the new helper(s)?

**Comment by ZeroIntensity** [comment]
FWIW, Bénédikt is on vacation right now. I don't think he'll be able to implement anything until next week.

File: Include/internal/pycore_pyerrors.h
Comment (vstinner) [inline Include/internal/pycore_pyerrors.h]:
```suggestion
 * Similar to _PyErr_SetString but decodes 'string' from the current locale.
```

**Review by vstinner — COMMENTED** [review]
I don't see the purpose of these internal functions since they are now used. This PR only adds dead code. I would prefer to see how these functions are used.

If only dlerror() uses it, why not using code using dlerror()?

**Comment by encukou** [comment]
> Bénédikt is on vacation right now

Ah! Thanks for the info. There's no rush :)

I converted this to draft, as the current version doesn't need reviews.

**Comment by picnixz** [comment]
I can include the usage of this new API in this PR if you prefer (my idea was to postpone it for a follow-up PR) but as Peter said, I don't have access to my laptop! So I'll do it in 10 days or so.

**Comment by ZeroIntensity** [comment]
I wrote a PR on my phone once 😉

**Comment by picnixz** [comment]
FTR I was able to use the new helpers in a few places but I made them return -1 if there is a decoding failure. Why? because some places explicitly ignore those decoding errors and fall back to generic errors when possible. 

However, I decided to expose them as part of the C API and it might happen that people may want the decoding error to have priority over the set exception (but I don't know how many). 

Since the caller expects an exception to be set, we shouldn't automatically clear the decoding error, and having an additional parameter that would store whether the decoding failed or not would likely be cumbersome (it would have been great to have overloading in C...)

There are some calls that I marked as "to check" because I don't know whether they were locale-dependent or not so I'll check them upon my return as well. For now, the dlerror() function is the only one I can remember as being locale-dependent (the `gdbm_strerr` only says "English text" but I'm not sure if it's because the man page is in English or not).

**Comment by picnixz** [comment]
Ok, I confirmed that `gdbm_strerror` indeed returns a localized string: https://git.gnu.org.ua/gdbm.git/tree/src/gdbmerrno.c#n155. So the little helper I wrote is indeed needed.

**Comment by picnixz** [comment]
I'll make it private for now and we can see in the future if we want or not to make it public.

> AFAIK, the expected failures should be something like MemoryError, a broken codec, or a KeyboardInterrupt. These should not be ignored

Then, what should we do with the code that currently ignores them?

File: Doc/data/refcounts.dat
Comment (vstinner) [inline Doc/data/refcounts.dat]: I'm not sure that it's useful to document the refcount behavior of internal functions. Other functions prefixed by _Py are part of the public C API and/or documented.

File: Include/internal/pycore_pyerrors.h
Comment (vstinner) [inline Include/internal/pycore_pyerrors.h]:
```suggestion
 * Exported for '_ctypes' shared extensions.
```

**Review by vstinner — COMMENTED** [review]
Is there a practical way to get a non-ASCII error messages not encoded to UTF-8 for the following functions?

* dlerror()
* gdbm_strerror()

How can I test (manually) modified code paths?

**Comment by picnixz** [comment]
> Is there a practical way to get a non-ASCII error messages not encoded to UTF-8 for the following functions?

I don't know :( Maybe changing the locale env. var. would be sufficient (assuming you have the translated messages) butI'm really bad at setting up locales so I don't really know =/

File: Doc/data/refcounts.dat
Comment (picnixz) [reply Doc/data/refcounts.dat]: I don't think it hurts (or we can have a separate file for that). My motivation is to move towards a maintained `refcounts.dat` for every function, even those that are not part of the Stable ABI, but if you think it's better to do it later, I can remove that entry.

File: Include/internal/pycore_pyerrors.h
Comment (picnixz) [reply Include/internal/pycore_pyerrors.h]: Oh, so "_ctypes" shared extensions means other shared extensions?

File: Doc/data/refcounts.dat
Comment (picnixz) [reply Doc/data/refcounts.dat]: I've decided to remove the entry so that we can add it in one go later.

File: Include/internal/pycore_pyerrors.h
Comment (vstinner) [inline Include/internal/pycore_pyerrors.h]:
```suggestion
 * Set an exception with the error message decoded from the current locale
 * encoding (LC_CTYPE).
```

**Review by vstinner — APPROVED** [review]
The change LGTM, but I didn't test it (manually).

**Review by serhiy-storchaka — COMMENTED** [review]
Translated error messages most likely depend on the LC_MESSAGE environment variable. It is a separate difficult issue if its encoding differs from LC_CTYPE's encoding.

But if it is non-ASCII because it includes file paths, we should use the FS encoding.

**Comment by picnixz** [comment]
> It is a separate difficult issue if its encoding differs from LC_CTYPE's encoding.

What should we do for that? should we fix this in a follow-up PR?

File: Include/internal/pycore_pyerrors.h
Comment (picnixz) [inline Include/internal/pycore_pyerrors.h]:
```suggestion
 * Exceptions occurring in decoding take priority over the desired exception.
 * In those cases, this function returns -1. Otherwise this returns 0 if the
 * localized exception has been successfully set.
```

**Comment by vstinner** [comment]
> Translated error messages most likely depend on the LC_MESSAGE environment variable. It is a separate difficult issue if its encoding differs from LC_CTYPE's encoding.

IMO it's reasonable to expect LC_MESSAGE and LC_CTYPE to use the same encoding in this API. It's already an enhancement compared to the current code which always expect UTF-8.

**Comment by encukou** [comment]
> Then, what should we do with the code that currently ignores them?

I should have caught that [when they were switched to `"surrogateescape"`](https://github.com/grgalex/cpython/commit/d179377df3a53b114101ae749d12f24aa4cd1fd9). Hopefully it's not too late now.

**Comment by picnixz** [comment]
Oh ok! I'll change it in a few hours (coming back home right now)

**Review by serhiy-storchaka — COMMENTED** [review]
Could you please add tests?

**Comment by picnixz** [comment]
> Could you please add tests?

Do you want tests for the additional messages or tests for the new function itself? For the first case, I don't really know how to do it since we need to check for the presence of the translation files but I can do it for the second.

**Comment by picnixz** [comment]
I won't be able to work for the rest of the week so I'll fix the CI failures this week-end-

File: Lib/test/test_ctypes/test_dlerror.py
Comment (vstinner) [inline Lib/test/test_ctypes/test_dlerror.py]: @serhiy-storchaka: We don't control dlerror() error message and translated error message. The test fails if translations are not installed or if the error message changes.

I'm not sure if it's a good idea to rely on the exact dlopen() translated error message.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [inline Lib/test/test_ctypes/test_dlerror.py]: This is not a C API test.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [inline Lib/test/test_ctypes/test_dlerror.py]: I think that `run_with_locale()` instead of `run_with_locales()` can be enough. But try to use non-UTF8 locales first. Try to use several locales, not only French. Add `''` at the end to use the current locale as a fallback.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [reply Lib/test/test_ctypes/test_dlerror.py]: We should not rely on the dlopen() error message. We only need to check that the error is not a UnicodeDecodeError.

Try to use a UTF8-undecodable bytes name, e.g. `b'\xff'` (bytes are accepted, right?).

File: Lib/test/test_dbm_gnu.py
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_gnu.py]: We should not rely on the error message. We only need to check that the error is not a UnicodeDecodeError.

File: Lib/test/test_dbm_gnu.py
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_gnu.py]: filename is TESTFN. It can be not encodable with the tested locale (we should try non-UTF8 locales first). Try to use ASCII-only name like "nonexisting". Try to use a UTF8-undecodable bytes name if bytes is accepted, otherwise use a name containing a surrogateescaped byte.

**Review by serhiy-storchaka — COMMENTED** [review]
I do not think that we need a C API test. It is just a private helper. In any case, the C API test is not working.

File: Lib/test/test_dbm_gnu.py
Comment (picnixz) [reply Lib/test/test_dbm_gnu.py]: Unfortunately, non-existing ASCII name wouldn't work because the error would be an errno error instead of a `gdbm.error`. I really need a specific kind of error to trigger in order to hit the translated messages. But I can't create an empty file with UTF8-undecodable byte names I think (os.open() fails with a ValueError). So I'm not entirely sure how I can do it =/ I'll try to find other ways offline (I will be offline until Friday)

Comment (serhiy-storchaka) [reply Lib/test/test_dbm_gnu.py]: `open(b'\xff')`

Comment (picnixz) [reply Lib/test/test_dbm_gnu.py]: How come `create_empty_file` failed when I tried? I'll need to check it later. But thanks! (I remember trying using the FN_NONDECODABLE file but didn't try with just '\xff').

Comment (vstinner) [reply Lib/test/test_dbm_gnu.py]: Why not testing an existing empty file with an ASCII filename? Something like:

```py
filename = "empty"
open(filename, "wb").close()
self.addCleanup(os_helper.unlink, filename)
# ... use filename ...
```

File: Lib/test/test_ctypes/test_dlerror.py
Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py]: I've chosen fr_FR.iso88591, ja_JP.sjis and zh_CN.gbk. Do you think others should be used?

File: Lib/test/test_ctypes/test_dlerror.py
Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py]: I've renamed it to `TestLocalization` (or something like that).

File: Lib/test/test_ctypes/test_dlerror.py
Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py]: Yes, and since I don't need to create it, I don't encounter issues with such filename. Depending on whether the locale is ISO-88591 or not, `\xff` decodes differently so it's fine (note that the OSError being raised does not have the error message as bytes but a string, so I can't check against bytes directly (I think?)).

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [inline Lib/test/test_ctypes/test_dlerror.py]: Why does it require Linux? It only should require `_ctypes.dlopen`.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [inline Lib/test/test_ctypes/test_dlerror.py]: Add also UTF-8 locales.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [inline Lib/test/test_ctypes/test_dlerror.py]: This is not always correct for other 8-bit locales. `b'\xff'` is not always decoded to `'\xff'`.

File: Lib/test/test_dbm_gnu.py
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_gnu.py]: TESTFN is non-ASCII which is encodable with the filesystem encoding when tests are imported. It may be not encodable in the specified test locale. So I suggest to use name containing `'\udcff'` or something like.

It may be impossible to create a non-UTF-8 file name on macOS, so catch an error in `create_empty_file()` and skip the test.

You may need to test this on Windows manually to ensure that the test works and not always skipped (if gdbm is available on Windows).

File: Modules/_ctypes/_ctypes.c
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:987]: Can this be tested too?

File: Modules/_ctypes/_ctypes.c
Comment (serhiy-storchaka) [inline Modules/_ctypes/_ctypes.c:3804]: Can this be tested too?

File: Modules/main.c
Comment (serhiy-storchaka) [inline Modules/main.c:377]: Then use `PyUnicode_DecodeLocale()` and add a test.

Other idea: use `str()` of the current error (before clearing it). It is not necessary `OSError`, so `errno` and `strerror()` may be wrong.

I do not think there is a simple way to test this, besides using embedded Python. So it is okay to not add a test for this.

File: Lib/test/test_dbm_gnu.py
Comment (picnixz) [reply Lib/test/test_dbm_gnu.py]: Ah I thought it would have been file. My question: since I don't care about the filename in this case (the error doesn't report it), I can just use an ASCII name and that's it right? (I'm just interested in not having UnicodeDecodeError)

Comment (serhiy-storchaka) [reply Lib/test/test_dbm_gnu.py]: Yes, you can just use an ASCII name.

File: Modules/_ctypes/_ctypes.c
Comment (picnixz) [reply Modules/_ctypes/_ctypes.c:3804]: I've added a test. Actually it was implicitly tested in `TestNullDlsym`, but only with the current locale.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [inline Lib/test/test_ctypes/test_dlerror.py]: Could not we use an existing library for these tests? They should not require gcc.

See for example test_loading.

Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py]: Yes, we can do it.

File: Modules/main.c
Comment (picnixz) [reply Modules/main.c:377]: Do you want me to actually address this in this PR or in a follow-up PR?

File: Lib/test/test_ctypes/test_dlerror.py
Comment (vstinner) [inline Lib/test/test_ctypes/test_dlerror.py]:
```suggestion
            dll.this_name_doesnt_exist
```

File: Lib/test/test_ctypes/test_dlerror.py
Comment (vstinner) [inline Lib/test/test_ctypes/test_dlerror.py]:
```suggestion
            c_int.in_dll(dll, 'this_name_doesnt_exist')
```

File: Lib/test/test_ctypes/test_dlerror.py
Comment (vstinner) [inline Lib/test/test_ctypes/test_dlerror.py:171]: Does test_null_dlsym() need the same decorators? PyPy doesn't seem to have _ctypes.dlsym() for example.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (vstinner) [inline Lib/test/test_ctypes/test_dlerror.py]:
```suggestion
            _ctypes.dlsym(dll, 'this_name_doesnt_exist')
```

File: Lib/test/test_ctypes/test_dlerror.py
Comment (vstinner) [inline Lib/test/test_ctypes/test_dlerror.py]: You may rename if to `libc_filename`.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:171]: Err... I don't know. For safety, I'll add those decorators as well.

File: Modules/main.c
Comment (serhiy-storchaka) [reply Modules/main.c:377]: It is up to you.

**Review by vstinner — APPROVED** [review]
LGTM

File: Lib/test/test_ctypes/test_dlerror.py
Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:171]: FTR, I decided to only skip the calls that use `dlopen` and `dlsym` but keep the calls that use `CDLL` and `c_int.in_dll` instead of skipping the entire test.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (encukou) [inline Lib/test/test_ctypes/test_dlerror.py:126]: Do you want `run_with_locales`, the decorator?
`run_with_locale` is a context manager.

Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:126]: OH! we should fix this then: https://docs.python.org/3/library/test.html#test.support.run_with_locale. And I wanted the decorator. Hopefully it will work...

Comment (picnixz) [inline Lib/test/test_ctypes/test_dlerror.py:126]:
```suggestion
        return test.support.run_with_locales(
```

File: Lib/test/test_dbm_gnu.py
Comment (picnixz) [inline Lib/test/test_dbm_gnu.py:209]:
```suggestion
    @support.run_with_locales(
```

File: Lib/test/test_ctypes/test_dlerror.py
Comment (encukou) [reply Lib/test/test_ctypes/test_dlerror.py:126]: Oh, never mind, it's both. Sorry for the noise.

Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:126]: I've actually changed the test so that all locales are tested and not only the first locale that could be found. Now, the tests are much faster to run so it's probably fine.

Comment (encukou) [reply Lib/test/test_ctypes/test_dlerror.py:126]: Hm, I'm not sure what the intended difference is. But, `run_with_locales` uses `subTest`, so at least it gives better error reporting.

Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:126]: - `run_with_locale` only uses the first locale that is correct
- `run_with_locales` runs for all specified locales (and skip those that cannot be found)
-

**Review by encukou — APPROVED** [review]

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [reply Lib/test/test_ctypes/test_dlerror.py:126]: It is both a decorator and a context manager.

The difference is that `run_with_locales` runs the test with all locales and can only be used as a decorator, `run_with_locale` only runs the test with the first available locale and can also be used as a context manager.

In this case it is enough to use `run_with_locale`.

Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:126]: Oh.. so you still want to use `run_with_locale`. Mmh, ok.

File: Lib/test/test_ctypes/test_dlerror.py
Comment (picnixz) [inline Lib/test/test_ctypes/test_dlerror.py]:
```suggestion
        return test.support.run_with_locale(
```

File: Lib/test/test_dbm_gnu.py
Comment (picnixz) [inline Lib/test/test_dbm_gnu.py]:
```suggestion
    @support.run_with_locale(
```

File: Lib/test/test_ctypes/test_dlerror.py
Comment (serhiy-storchaka) [reply Lib/test/test_ctypes/test_dlerror.py:126]: Most tests use `run_with_locale`. I added `run_with_locales` recently specially for tests which should be run for several locales.

Perhaps in these cases we should use a different approach and continue to run the test for different locales until we find one with translations, and mark the test skipped if no such locale is found. I'll think about this idea if I have a chance. For now, you can use what you prefer.

Comment (picnixz) [reply Lib/test/test_ctypes/test_dlerror.py:126]: (I'll keep `run_with_locale` since I don't want yet another edit and since it's probably fine to only test a single locale for those specific tests)

**Comment by serhiy-storchaka** [comment]
This is a bug fix. It should be backported.

**Comment by picnixz** [comment]
Ah! I think we should also have a NEWS entry. I'll add one. Sorry for forgetting about it :(

**Comment by picnixz** [comment]
Concerning the failure on 3.12 backport it's because we did not backported the PR on `dlerror()` usages yet. I've opened a PR for that #127764

**Comment by picnixz** [comment]
Grrr. I thought that the conflicts would be easily solved but apparently not. I'll do the backport manually.

**Comment by picnixz** [comment]
Ok, so I didn't expect that on some archs, the error message would be different...

**Comment by vstinner** [comment]
Tests are failing on PPC64LE RHEL8 LTO 3.x:

> AssertionError: 'libc.so.6' not found in '/lib64/glibc-hwcaps/power9/libc-2.28.so: undefined symbol: this_name_does_not_exist'

**Pull request merged**
